### ISSUE TITLE
ubipop sometimes gets stuck in as_completed()

### ISSUE THREAD

**Issue by rbikar** [release-engineering/ubi-population-tool#49]
Ubipop sometimes gets stuck in as_completed() while waiting for result of future.

Recently it happened in that place:
ubipop/__init__.py
```python
                name = package_pattern.name
                arch = None if package_pattern.arch in ('*', None) else package_pattern.arch
                fts[(self._executor.submit(self.pulp.search_rpms,
                                          repo, name, arch))] = name

214>       for ft in as_completed(fts):
                packages = ft.result()
                if packages:
                   packages_dict[fts[ft]].extend(packages)
```

In depth, it was stuck in:
python3.7/threading.py
```python
 291            self._waiters.append(waiter)
 292            saved_state = self._release_save()
 293            gotit = False
 294            try:    # restore state no matter what (e.g., KeyboardInterrupt)
 295                if timeout is None:
>296                    waiter.acquire()
 297                    gotit = True
 298                else:
 299                    if timeout > 0:
 300                        gotit = waiter.acquire(True, timeout)
 301                    else: 
```
I'm not sure what is exact reason for that but we can probably add some reasonable timeout to futures.

**Comment by rohanpm** [comment]
If anyone manages to catch a process hanging in this state, it would be useful to get backtraces from *all* threads, e.g. with gdb:

`gdb --pid <pid> --batch --eval 'thread apply all py-bt' 2>&1 | tee backtrace.txt`

**Comment by rbikar** [comment]
managed to get backtraces from all threads:
[backtrace.txt](https://github.com/release-engineering/ubi-population-tool/files/3093207/backtrace.txt)

**Comment by rbikar** [comment]
Also this seems to happen only when jenkins job is triggered manually.

**Comment by rohanpm** [comment]
That's great.  To summarize:

- Thread 1: that's the main thread, waiting on as_completed
- Thread 2: thread for RetryExecutor, waiting for any futures to be retried.  It's expected to have a thread in this state.
- Threads 3-6: all of them are in the middle of an HTTP request to Pulp - all blocked on read.

Since threads 3-6 are actually doing something, it doesn't look like there's any bug with respect to how ubipop is setting up or waiting for the futures.

A couple of theories:

- those threads 3-6 aren't really hung but instead the requests are very slow, and it looks like a hang if you don't wait long enough - especially with the silent retries added on top
- it's a race condition caused by sharing the same requests session by all the threads (issue 57)

Fixing issue 57, adding more logging and adding timeouts all seem like good ideas here.

**Comment by alexandrevicenzi** [comment]
@rohanpm a fix for #57 is on the way.

About logging I think we could add logs for retries, so we know that it's not stuck, it's just retrying. Based on our retry count and back off a request could be delayed for almost 10 min, supposing it fails 9 times.

Can you clarify where we should add timeouts?

**Comment by rohanpm** [comment]
For logging, I can think of two ways to add that.

Firstly, we could add try/log/raise in the code which fires requests to Pulp, e.g.

```
try:
  return self._pulp.some_method(...)
except:
  LOG.warn("Pulp request failed", exc_info=1)
  raise
```

That's nice and generic, but the downside is that this code doesn't know whether the request is about to be retried.  Some might see the "Pulp request failed" log messages and assume ubipop didn't finish its work.

Otherwise, installing a custom [RetryPolicy](https://rohanpm.github.io/more-executors/reference/retry.html#more_executors.retry.RetryPolicy) which logs in `should_retry` before returning True or False is possible.  Then the logging code knows for sure whether the failed request will be retried.  The downside is that it requires more deeply integrating with more-executors library.

I'd prefer the first approach.

For timeouts, as_completed should probably have a configurable timeout with large default - e.g. 8 hours if we expect an ubipop job would never reasonably take that long.  The idea there would be just to change the failure mode from "ubipop was running all night and didn't do anything" to "there was a slow, failed ubipop job overnight, but it eventually recovered and started a new run".

**Comment by rohanpm** [comment]
In any case, since this issue could drag on forever until we have "proof" that it no longer gets stuck - I'd suggest we just close it in favor of #57 and only reopen if we still observe it getting stuck after that.

**Comment by alexandrevicenzi** [comment]
@rohanpm I think that the logging would help in the future to debug if it happens again. Just closing this with no changes it's not a good idea IMHO.

**Comment by rohanpm** [comment]
@alexandrevicenzi yes, my point is just that we can't really ever prove/demonstrate that "ubipop sometimes gets stuck in as_completed()" is fixed, so the issue could drag on forever if we're not aggressive about closing it.  Logging isn't going to fix that either, so I'd have thought it makes more sense to file a separate issue for logging improvements.

**Comment by alexandrevicenzi** [comment]
@rbikar what are your thoughts? Should we close this?

**Comment by rbikar** [comment]
@alexandrevicenzi 
I think that Rohan's proposal is ok. Close that for now, see if it will happen again with #57 landed and file issue for logging upgrade.
